### ISSUE TITLE
folder.getItem() and folder.files.addChuncked() fail for libraries with >5000 folders

### ISSUE THREAD

**Issue by LastGunslinger** [pnp/pnpjs#2695]
### Category
- [ ] Enhancement
- [X] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [  3.15.0  ]

Please specify what version(s) of SharePoint you are targeting: [ SharePoint Online ]

I have an SPFx app that creates document sets for users and allows them to upload files to that document set at the same time. With over 20,000 combined document sets and items in the library it has been working fine, but after the library hit over 5000 document sets(folders) in the library, users started reporting that the app was showing the Microsoft.SharePoint.SPQueryThrottledException when trying to create new document sets(essentially the same as #1528). 

I was able to narrow down the issue to two areas. After the app creates a folder, I then get use the returned object to get the list item associated with that folder and update the metadata. This is pseudocode, but this is what I am doing:

```javascript
const newFolder = await spfi.web.lists.getById(listGuid).rootFolder.addSubfolderUsingPath(folderName)

// The folder is created correctly, but this line causes a throttle error if there are >5000 folders in the library
const newItem = await newFolder.getItem()
```

The second area where I was seeing this error is another folder operation, so I believe they are linked. After a user creates a folder, I upload their selected files, and that is where the next error occurs:

```javascript
// This line also fails with a throttle error BEFORE the file upload starts
newFolder.files.addChunked(
    fileName,
    fileContent,
    (data) => Logger.log({level: LogLevel.Info, message: `Upload progress for ${fileName}`, data}),
    false
)
```

This is a pretty crucial app for my users, and I had to disable the file upload feature because of this bug. I can technically work around the first issue by using pagination and querying for an item matching the FileLeafRef of the folder that I just created, but this is slower than just getting the item directly from the folder object. From what I can see, issue #1528 was closed to inactivity - is this an indication that this can't or won't be fixed?

**Comment by LastGunslinger** [comment]
@bcameron1231 I appreciate someone looking into this, but I want to respectfully push back against the tagging of this issue as a question. This *is* a bug, and my users are not able to upload files using PnPJS in this document library because of this. This is perhaps my fault for ending the above with a question, but I think this does rightfully qualify as a bug - something that should work is not working. The other linked issue was closed without resolution, so in opening this I was hoping that this would be acknowledged as a legitimate issue and either investigated and patched in a future release or closed as a could not / would not fix, especially for other users who may run into the same issue.

**Comment by bcameron1231** [comment]
I was able to do some testing and I have good and bad news unfortunately. 

So unfortunately, the APIs we use to get the item, or to add the files from the folder object will not work with the 5,000 item limit. The reason seems to be because it has to do a query to the folder first, lists the files and add a new one. Which I'm sure you've seen in the calls... /folders('name')/files/Add(....). It's the /files on the SharePoint REST API that I believe is throwing the error when you're over the 5,000 item limit. 

The method that would seemingly only work for uploading the file is to reference everything by the path.  Please see below. I add the subfolder named "Beau1". Then I addChucked, but instead of just the filename, I use the path to folder and the filename. I know it might not be ideal, but I don't think we have a way to fix the other APIs that do the add on the folder objects themselves.

```
  const newFolder = await sp.web.lists.getByTitle("Documents").rootFolder.addSubFolderUsingPath("Beau1");
  
   sp.web.lists.getByTitle("Test Documents").rootFolder.files.addChunked("/sites/SiteName/Shared Documents/Beau1/Test.txt",blob,(data)=>console.log("test"));
```
            
I hope at this minimum that this can unblock you and you continue to use your application.

**Comment by LastGunslinger** [comment]
@bcameron1231, I want to thank you again for investigating this for me. I found something interesting while I was testing that I wanted to share for you and anyone else who might stumble on this issue. I feel that I don't know enough about the nuts and bolts of SharePoint's API to understand why it works this way, but perhaps smarter people do.

I recreated my initial issue, then switched to using your suggestion above for adding files, and it did work. That was good news, but only fixed one of my issues, uploading files. The other issue, not being able to fetch a folder's IItem object, was still using a hacky workaround where I was querying the list for recently created files and filtering them based on the FileLeafRef property - not a good solution at all. I replaced it with the code below, which seems to be working for now:

```javascript
const rootFolder = spfi.web.lists.getById(listGuid).rootFolder
await rootFolder.addSubfolderUsingPath(folderName)

// Get the root folder info so that I can use its ServerRelativeUrl property since this same app used on many sites
const rootFolderInfo = await rootFolder()

// Fetch the newly created folder directly from the web object
const newFolder = spfi.web.getFolderByServerRelativePath(`${rootFolderInfo.ServerRelativeUrl}/${folderName}`)

// This works now if I fetch the folder using the full server-relative URL above
const newItem = await newFolder.getItem()

// Also, this works
newFolder.files.addChunked(
    fileName,
    fileContent,
    (data) => Logger.log({level: LogLevel.Info, message: `Upload progress for ${fileName}`, data}),
    false
)
```

This should work to allow my users to continue uploading files. I appreciate you taking the time, and believe me, if it were possible for me to re-architect this library in a different way, I would have already done so.

On another note, I just wanted to let you know that I've been using this library for years, and this is really the only pain point I've had besides migrating from v2 to v3. Thanks again for all the work you and the team are doing.
